### ISSUE TITLE
Cannot connect to UDM / UDM-Pro

### ISSUE THREAD

**Issue by mgbowman** [mgbowman/openhab2-addons#6]
Discussed here:

https://community.openhab.org/t/ubiquiti-unifi-binding-feature-discussion/14520/1013?u=mgbowman

**Comment by mgbowman** [comment]
I use [Postman](https://www.postman.com) - a collaborative API platform I recently discovered that's helped me on this project and others. I'm thinking it's the best to exchange debug data and try and figure out what's going on!

I have a collection of UniFi API calls that we can share and exchange data.

**Comment by skrebber** [comment]
Does this help?
https://github.com/Art-of-WiFi/UniFi-API-client
there is a version, which is capable of the changed API of the UDM/UDM-Pro

**Comment by mgbowman** [comment]
@skrebber thanks for the link!

I will give it a look and see what I can do.

**Comment by neoice** [comment]
I was successfully using the OpenHAB binding with the `linuxserver/unifi-controller` Docker container successfully. however, the UDM / UDM-PRO devices appear to use a different authentication flow and I suspect that breaks the binding with the following error:

```
[WARN ] [.handler.UniFiControllerThingHandler] - Unhandled exception while refreshing the UniFi Controller unifi:controller:home - com.google.gson.stream.MalformedJsonException: Use JsonReader.setLenient(true) to accept malformed JSON at line 2 column 12 path $
```

I even tried the latest build (`#118`) from your Jenkins => https://jenkins.otr.mx/job/openhab2-unifi-binding/

**Comment by neoice** [comment]
this is still relevant to my interests. do you need help accessing or obtaining a UDM-Pro...? I can definitely run whatever debugging or troubleshooting or alpha code you want or possible just ship you one? there's no direct messages on Github in 2020 right? you can probably guess my email address.

**Comment by lexusburn** [comment]
I had a look at the implementation for UDM on https://github.com/Art-of-WiFi/UniFi-API-client/blob/master/src/Client.php and it seems that ubiquiti hat done some changes for their new unifi-os based devices:

- We don't have to use a specified port for api calls anymore (no 8443)
- auth route is '/api/auth/login' instead of '/api/login'
- all other routes to handle some network stuff had to be prefixed with '/proxy/network', so that "${host}/api/s/${site}/stat/sysinfo" will be "${host}/proxy/network/api/s/{site}/stat/sysinfo"
- all routes had to include the header named 'x-csrf-token' with payload 'csrfToken' from JWT cookie.

That seems to be all. I can use the api with postman with this changes.
But i don't have a working development environment at the moment to add this changes to the plugin. :-(

@mgbowman can you add this special workflows for new unifi-os based devices please? 
@neoice and i can test this changes if you give us access on a compiled jar.

**Comment by mgbowman** [comment]
@lexusburn Thanks for that!

I had to relocate and start rebuilding my UniFi infrastructure from scratch. I now have UDM and plan to fix the binding to work with both CKs and UDMs.

Stay tuned!

**Comment by mgbowman** [comment]
Can you guys give me some feedback on your setups?

 * Are you using UDM Base or UDM Pro? 
 * What UDM firmware version are you running?

**Comment by sosolid** [comment]
Hi Matthew,

UDM Pro running 1.8.0.2888

**Comment by mgbowman** [comment]
I forgot to mention that I have a UDM Base running v1.5.6 (latest base stable).

I'm wondering if there's going to be an issue with supporting 1.5 (Legacy) vs 1.7+ (UniFi OS) and what's the best way forward to make sure we can support both going forward.

Thoughts?

**Comment by mdillmann** [comment]
Hi Matthew,
I also run UDM Pro on 1.8.0.

I haven't followed the discussion why people would want to stay on legacy 1.5 for base. Unifi themselves would want all UDM (Base/PRO) to be on UniFI-OS going forward. So I personally would not focus on legacy.
Does legacy on UDM behave differently than Cloudkey/Controller ?

**Comment by skrebber** [comment]
Hi,
I'm running UDM Pro on 1.6.5
br

**Comment by lexusburn** [comment]
Hi,

I'm running UDM Pro on 1.6.5

@mdillmann: UDM had new OS which had some changes which old controllers don't get in future.

**Comment by mgbowman** [comment]
> I haven't followed the discussion why people would want to stay on legacy 1.5 for base.

1.7+ is consider BETA for UDM base 🙃 

> Does legacy on UDM behave differently than Cloudkey/Controller ?

That was my next question, what's the controller version reported by 1.7+ firmwares?

**Comment by mdillmann** [comment]
Well,
they ship Controller 5.14.22 with 1.8.0. For UDM-PRO they tagged it "stable" and "beta" for the UDM. But they recently dropped stable/beta tags and just name everything "official".

Some people manually install controller 6.x.

I gues they want to wait until the 6.x series of the controller is more mature and then release new FW for UDM and UDM-Pro

**Comment by mgbowman** [comment]
So mine is 5.12.60

Looks like we need test setups for the following:
 * 5.12.x (covered by me)
 * 5.14.x
 * 6.0.x

I will get the first 2.5 build with UDM base 1.5.x support posted soon. Hopefully it will work fine for the rest of your setups.

**Comment by skrebber** [comment]
I can test 5.12.67

**Comment by mdillmann** [comment]
> So mine is 5.12.60
> 
> Looks like we need test setups for the following:
> 
>     * 5.12.x (covered by me)
> 
>     * 5.14.x
> 
>     * 6.0.x
> 
> 
> I will get the first 2.5 build with UDM base 1.5.x support posted soon. Hopefully it will work fine for the rest of your setups.

I'll test 5.14.x with the Pro (UniFI-OS). As I also still have a OH3 dev environment, I could do tests/migration there once it stabilizes

**Comment by DaHarry** [comment]
Hi, i can test UDM-Pro 5.14.22. Where i can get the last dev Binding?

**Comment by DJManuel** [comment]
Hi, Has there been any progress lately? I have an UDM Pro with 6.0.27 that i could test the binding with, but as @DaHarry asked already: Where can we get the latest dev Binding?
I have at the moment OpenHAB 2.5.10 setup as well as OpenHAB 3.0.0M3 running but I'm rather new to OpenHub in general (only have it since a few weeks, haven't had a look at file configs / rule engine, as the main functionality at the moment is still done in NodeRED).

**Comment by rtfm87** [comment]
I would also test the DEV binding. I have a CloudKey Gen2 Plus with Unifi OS Firmware 2.0.22 and Unifi Network 6.0.41

**Comment by seaside1** [comment]
I have implemented the authentication for UniFi Protect Openhab binding for firmware 2.0.18 and above (UniFi OS), which works for cloudkey.
Take a look at the source code if it helps:

https://github.com/seaside1/unifiprotect/blob/main/src/main/java/org/openhab/binding/unifiprotect/internal/model/request/UniFiProtectLoginRequest.java

https://github.com/seaside1/unifiprotect/blob/f377e380447786576351613205f70d290b605c81/src/main/java/org/openhab/binding/unifiprotect/internal/model/UniFiProtectNvr.java#L105

**Comment by mdillmann** [comment]
Firmware 1.8.4 (and lately 1.8.5 bugfix) are available for Pro and base. Both contain controller version 6.
I plan on upgrading soon so I could test this version.

@mgbowman: did you find any time to work on this binding ?

**Comment by schinken** [comment]
Is there any info I can provide to get this working again? I own an UDM Pro with version 1.8.5. If you need any details, I can look into it. 

Thanks for your awesome binding btw! :)

**Comment by tjarb** [comment]
Hi all,

Just curious because I'be upgraded my network from USG3 to UDM-pro and also facing issues with the integrated unifi network controller (6.5.55) and the unifi binding.

Is there a solution?

Best regards,

**Comment by schinken** [comment]
Yes: Use the latest OpenHAB release and tick the check on "Unifi OS" :)

**Comment by tjarb** [comment]
Hi @schinken ,

Thank you for your reply.
I did check the unifi os option but it does not work. I'm using the credentials as used to login the unifi.ui cloud service (worked for the cloud key).

I didn't set any port to connect to (seems to default to 8443 in the configuration).

Best regards,

**Comment by schinken** [comment]
Hi @tjarb,

i've created a separate user locally on the UDM

![image](https://user-images.githubusercontent.com/187176/146653811-f143d486-e9ac-4c83-bb85-2e753ce6307b.png)

Works like a charm :)

**Comment by tjarb** [comment]
Hi @schinken,

I also tries that one but doesn't work either.

` 
UID: unifi:controller:UDMpro
label: UniFi Controller
thingTypeUID: unifi:controller
configuration:
  host: 192.168.2.254
  refresh: 10
  unifios: true
  password: Password123
  port: 8443
  username: openhab
`
Throws a configuration error 

Best regards,

**Comment by schinken** [comment]
hm okay... the only difference I have is, that my controller listens to :443 as port, not 8443 - Maybe your's, too?

**Comment by tjarb** [comment]
Hi @schinken ,

Yes thats it, it's listening on port 443!
Thank you for the info, should this be added to the binding page?

Best regards,

**Comment by schinken** [comment]
Indeed it should. UDM is different from the software controller before, i guess :)

**Comment by schinken** [comment]
@tjarb  https://github.com/openhab/openhab-addons/pull/11812

**Comment by tjarb** [comment]
Ah yes Great,

Thank you for the assistance.and the follow-up.
